chain/db/statedb.go

```diff
@@ -202,32 +288,98 @@ func (sdb *StateDB) deleteAccount(addr common.Uint160, acc *account) error {
 	return nil
 }
 
-func (sdb *StateDB) Finalise(deleteEmptyObjects bool) {
+func (sdb *StateDB) updateNanoPay(id string, nanoPay *nanoPay) error {
+	buff := bytes.NewBuffer(nil)
+	err := nanoPay.Serialize(buff)
+	if err != nil {
+		panic(fmt.Errorf("can't encode nano pay %v: %v", nanoPay, err))
+	}
+
+	return sdb.trie.TryUpdate(append(NanoPayPrefix, id...), buff.Bytes())
+}
+
+func (sdb *StateDB) deleteNanoPay(id string) error {
+	err := sdb.trie.TryDelete(append(NanoPayPrefix, id...))
+	if err != nil {
+		return err
+	}
+
+	delete(sdb.nanoPay, id)
+	return nil
+}
+
+func (sdb *StateDB) updateNanoPayCleanup(height uint32, nanoPayCleanup map[string]struct{}) error {
+	buff := bytes.NewBuffer(nil)
+
+	if err := serialization.WriteVarUint(buff, uint64(len(nanoPayCleanup))); err != nil {
+		panic(fmt.Errorf("can't encode nano pay cleanup %v: %v", nanoPayCleanup, err))
+	}
+	npcs := make([]string, 0)
+	for id := range nanoPayCleanup {
+		npcs = append(npcs, id)
+	}
+	sort.Strings(npcs)
+	for _, id := range npcs {
+		if err := serialization.WriteVarString(buff, id); err != nil {
+			panic(fmt.Errorf("can't encode nano pay cleanup %v: %v", nanoPayCleanup, err))
+		}
+	}
+
+	return sdb.trie.TryUpdate(append(NanoPayPrefix, getNanoPayCleanupId(height)...), buff.Bytes())
+}
+
+func (sdb *StateDB) deleteNanoPayCleanup(height uint32) error {
+	err := sdb.trie.TryDelete(append(NanoPayPrefix, getNanoPayCleanupId(height)...))
+	if err != nil {
+		return err
+	}
+
+	delete(sdb.nanoPayCleanup, height)
+	return nil
+}
+
+func (sdb *StateDB) Finalise() {
 	sdb.accounts.Range(func(key, v interface{}) bool {
 		if addr, ok := key.(common.Uint160); ok {
 			if acc, ok := v.(*account); ok {
-				if deleteEmptyObjects && acc.Empty() {
-					sdb.deleteAccount(addr, acc)
+				if acc.Empty() {
+					sdb.deleteAccount(addr)
 				} else {
 					sdb.updateAccount(addr, acc)
 				}
 			}
 		}
 		return true
 	})
+
+	for id, nanoPay := range sdb.nanoPay {
+		if nanoPay == nil {
+			sdb.deleteNanoPay(id)
+		} else {
+			sdb.updateNanoPay(id, nanoPay)
+		}
+	}
+
+	for height, nanoPayCleanup := range sdb.nanoPayCleanup {
+		if nanoPayCleanup == nil {
+			sdb.deleteNanoPayCleanup(height)
+		} else {
+			sdb.updateNanoPayCleanup(height, nanoPayCleanup)
+		}
+	}
 }
 
-func (sdb *StateDB) IntermediateRoot(deleteEmptyObjects bool) common.Uint256 {
-	sdb.Finalise(deleteEmptyObjects)
+func (sdb *StateDB) IntermediateRoot() common.Uint256 {
+	sdb.Finalise()
 	return sdb.trie.Hash()
 }
 
-func (sdb *StateDB) CommitTo(deleteEmptyObjects bool) (root common.Uint256, err error) {
+func (sdb *StateDB) CommitTo() (root common.Uint256, err error) {
```

Comment:
@NoelBright I think we can merge Finalise() and CommitTo() later (not in this PR), quite a bit redundant code.

Reply:
merged